org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/InvisibleProjectImporter.java

```diff
@@ -167,6 +177,141 @@ public static boolean loadInvisibleProject(IPath javaFile, IPath rootPath, boole
 		return true;
 	}
 
+	/**
+	 * Based on the trigger file and its belonging source folder, search more places and collect
+	 * the valid source paths.
+	 * @param triggerFilePath the path of the import trigger file.
+	 * @param triggerFolder the folder which contains the trigger file. A search will be executed
+	 *                      based on the trigger folder.
+	 * @param linkedFolder the invisible project's linked folder.
+	 */
+	private static Collection<IPath> collectSourcePaths(IPath triggerFilePath, IFolder triggerFolder,
+			IFolder linkedFolder) {
+		IPath linkedFolderPath = linkedFolder.getLocation();
+		Collection<File> foldersToSearch = collectFoldersToSearch(triggerFilePath, triggerFolder, linkedFolderPath);
+		IProject currentProject = linkedFolder.getProject();
+		if (currentProject == null) {
+			return Collections.emptySet();
+		}
+		Collection<IPath> triggerFiles = collectTriggerFiles(currentProject, foldersToSearch);
+		
+		Set<IPath> sourcePaths = new HashSet<>();
+		sourcePaths.add(triggerFolder.getFullPath());
+
+		for (IPath javaFilePath : triggerFiles) {
+			String packageName = getPackageName(javaFilePath, linkedFolderPath);
+			IPath directory = inferSourceDirectory(javaFilePath.toFile().toPath(), packageName);
+			if (directory == null || !linkedFolderPath.isPrefixOf(directory)
+					|| isPartOfMatureProject(directory)) {
+				continue;
+			}
+			IPath relativeSourcePath = directory.makeRelativeTo(linkedFolderPath);
+			IPath sourcePath = linkedFolder.getFolder(relativeSourcePath).getFullPath();
+			sourcePaths.add(sourcePath);
+		}
+		return sourcePaths;
+	}
+
+	/**
+	 * Collect folders that may contain Java source files.
+	 * @param triggerFilePath the path of the import trigger file.
+	 * @param triggerFolder the folder which contains the trigger file. A search will be executed
+	 *                      based on the trigger folder.
+	 * @param linkedFolderPath the path of invisible project's linked folder.
+	 */
+	private static Collection<File> collectFoldersToSearch(IPath triggerFilePath, IFolder triggerFolder,
+			IPath linkedFolderPath) {
+		Set<File> foldersToSearch = new HashSet<>();
+		if (Objects.equals(triggerFolder.getLocation(), linkedFolderPath)) {
+			foldersToSearch.addAll(getDirectChildFolders(triggerFilePath, triggerFolder));
+		} else {
+			foldersToSearch.addAll(getSiblingFolders(triggerFolder, linkedFolderPath));
```

Comment:
We search for child folders when the trigger folder happens to be at the root of the project, and we search for sibling folders of the trigger folder otherwise. What about folders that exist betwwen the linked folder and the trigger folder ? Couldn't those be added ?

Reply:
Because the trigger folder here is source root that we inferred from the trigger file. I guess for a typical unmanaged folder project, it will be more likely to contain other source roots at the sibling locations.

But that's just my assumptions, we can add more scanning targets if the user's feedback turns out that this assumption is not enough.

Anyway, at least we can still recover the source roots increasingly as long as user opens the file that we missed.  :)

Reply:
Sounds good. I'll just have a last look, but I think we're basically ready to merge this.